lib/active_shipping/shipment_packer.rb

```diff
@@ -11,63 +11,90 @@ class ExcessPackageQuantity < StandardError; end
-      total_weight = 0
-      items.map!(&:symbolize_keys).each do |item|
-        total_weight += item[:quantity].to_i * item[:grams].to_i
-
-        if item[:grams].to_i > maximum_weight
-          raise OverweightItem, "The item with weight of #{item[:grams]}g is heavier than the allowable package weight of #{maximum_weight}g"
-        end
 
-        if total_weight > maximum_weight * EXCESS_PACKAGE_QUANTITY_THRESHOLD
-          raise ExcessPackageQuantity, "Unable to pack more than #{EXCESS_PACKAGE_QUANTITY_THRESHOLD} packages"
+    class << self
+      def pack(items, dimensions, maximum_weight, currency)
+        return [] if items.empty?
+        packages = []
+        items.map!(&:symbolize_keys)
+
+        # Naive in that it assumes weight is equally distributed across all items
+        # Should raise early enough in most cases
+        total_weight = calculate_total_weight(items, maximum_weight)
+        items = items.map(&:dup).sort_by! { |i| i[:grams].to_i }
+
+        state = :package_empty
+        while state != :packing_finished
+          case state
+          when :package_empty
+            package_weight, package_value = 0, 0
+            state = :filling_package
+          when :filling_package
+            items.each do |item|
+              quantity = determine_fillable_quantity_for_package(item, maximum_weight, package_weight)
+              package_weight += item_weight(quantity, item[:grams])
+              package_value += item_value(quantity, item[:price])
+              item[:quantity] = item[:quantity].to_i - quantity
+            end
+
+            items.reject! { |i| i[:quantity].to_i == 0 }
+            state = :package_full
+          when :package_full
+            packages << ActiveShipping::Package.new(package_weight, dimensions, :value => package_value, :currency => currency)
+            state = items.any? ? :package_empty : :packing_finished
+          end
         end
+
+        packages
       end
 
-      items = items.map(&:dup).sort_by! { |i| i[:grams].to_i }
-
-      state = :package_empty
-      while state != :packing_finished
-        case state
-        when :package_empty
-          package_weight, package_value = 0, 0
-          state = :filling_package
-        when :filling_package
-          items.each do |item|
-            quantity = if item[:grams].to_i <= 0
-              item[:quantity].to_i
-            else
-              # Grab the max amount of this item we can fit into this package
-              # Or, if there are fewer than the max for this item, put
-              # what is left into this package
-              [(maximum_weight - package_weight) / item[:grams].to_i, item[:quantity].to_i].min
-            end
+      private
 
-            item_weight = quantity * item[:grams].to_i
-            item_value = quantity * Package.cents_from(item[:price])
+      def calculate_total_weight(items, maximum_weight)
+        items.inject(0) do |total_weight, item|
```

Comment:
I've run some really basic benchmarks and the results were the following (`original` is the implementation I had, `suggestion` is with your suggested changes):

With a single item thats overweight:

```
.       user     system      total        real
original  0.000000   0.000000   0.000000 (  0.000008)
suggestion  0.000000   0.000000   0.000000 (  0.000011)
.
```

with large amount of items (1000), each having low weigh 1g, each low quantity of 10:

```
          user     system      total        real
original  0.000000   0.000000   0.000000 (  0.000320)
suggestion  0.000000   0.000000   0.000000 (  0.000359)
..
```

1000 items, low weight, high quantity

```
   .       user     system      total        real
original  0.000000   0.000000   0.000000 (  0.000339)
suggestion  0.000000   0.000000   0.000000 (  0.000359)
.
```

Seems to be almost the same, but original implementation performing slightly better. I'll keep it as is for now.